### ISSUE TITLE
Converted indicator pattern fails-hard on validator

### ISSUE THREAD

**Issue by emmanvg** [oasis-open/cti-stix-elevator#239]
While trying to convert a STIX 1.1.1 package the `stix2elevator` hangs in the validation step for the STIX 2.1 pattern being created. I managed to nailed it down to this call:

https://github.com/oasis-open/cti-stix-elevator/blob/8d8d061595a0a633c0a3cd30a8c43cfa3901f2ad/stix2elevator/convert_stix.py#L1248

and this is the pattern generated which looks odd:

`
[file:hashes.MD5 = 'd41d8cd98f00b204e9800998ecf8427e' AND ([file:name = 'TheFileName.exe' AND file:parent_directory_ref.path = '/usr/local'] AND [file:name = 'bin']) AND file:size = 123456]
`

Library versions:
```
stix 1.1.1.19
cybox 2.1.0.21
lxml 4.5.0
mixbox 1.0.4
stixmarx 1.0.7
stix2patterns 1.3.1
stix2validator 2.0.2
stix2elevator 3.0.0
pycountry 20.7.3
```

I suspect the problem is not on the validator and instead in the code that generates the pattern. The XML below can be used to re-create issue and the stacktrace follows:

``` xml
<stix:STIX_Package
	xmlns:stix-ciqidentity="http://stix.mitre.org/extensions/Identity#CIQIdentity3.0-1"
	xmlns:stixVocabs="http://stix.mitre.org/default_vocabularies-1"
	xmlns:AIS="http://www.us-cert.gov/STIXMarkingStructure#AISConsentMarking-2"
	xmlns:FileObj="http://cybox.mitre.org/objects#FileObject-2"
	xmlns:cyboxVocabs="http://cybox.mitre.org/default_vocabularies-2"
	xmlns:indicator="http://stix.mitre.org/Indicator-2"
	xmlns:marking="http://data-marking.mitre.org/Marking-1"
	xmlns:cybox="http://cybox.mitre.org/cybox-2"
	xmlns:cyboxCommon="http://cybox.mitre.org/common-2"
	xmlns:stixCommon="http://stix.mitre.org/common-1"
	xmlns:stix="http://stix.mitre.org/stix-1"
	xmlns:xal="urn:oasis:names:tc:ciq:xal:3"
	xmlns:xpil="urn:oasis:names:tc:ciq:xpil:3"
	xmlns:xnl="urn:oasis:names:tc:ciq:xnl:3"
	xmlns:CISA="http://www.us-cert.gov/nccic"
	xmlns:xlink="http://www.w3.org/1999/xlink"
	xmlns:ds="http://www.w3.org/2000/09/xmldsig#"
	xmlns:xs="http://www.w3.org/2001/XMLSchema"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"

	xsi:schemaLocation="http://stix.mitre.org/extensions/Identity#CIQIdentity3.0-1 http://stix.mitre.org/XMLSchema/extensions/identity/ciq_3.0/1.1.1/ciq_3.0_identity.xsd
	http://stix.mitre.org/default_vocabularies-1 http://stix.mitre.org/XMLSchema/default_vocabularies/1.1.1/stix_default_vocabularies.xsd
	http://www.us-cert.gov/STIXMarkingStructure#AISConsentMarking-2 http://www.us-cert.gov/sites/default/files/STIX_Namespace/AIS_Bundle_Marking_1.1.1_v1.0.xsd
	http://cybox.mitre.org/objects#FileObject-2 http://cybox.mitre.org/XMLSchema/objects/File/2.1/File_Object.xsd
	http://cybox.mitre.org/default_vocabularies-2 http://cybox.mitre.org/XMLSchema/default_vocabularies/2.1/cybox_default_vocabularies.xsd
	http://stix.mitre.org/Indicator-2 http://stix.mitre.org/XMLSchema/indicator/2.1.1/indicator.xsd
	http://data-marking.mitre.org/Marking-1 http://stix.mitre.org/XMLSchema/data_marking/1.1.1/data_marking.xsd
	http://cybox.mitre.org/cybox-2 http://cybox.mitre.org/XMLSchema/core/2.1/cybox_core.xsd
	http://cybox.mitre.org/common-2 http://cybox.mitre.org/XMLSchema/common/2.1/cybox_common.xsd
	http://stix.mitre.org/common-1 http://stix.mitre.org/XMLSchema/common/1.1.1/stix_common.xsd
	http://stix.mitre.org/stix-1 http://stix.mitre.org/XMLSchema/core/1.1.1/stix_core.xsd
	urn:oasis:names:tc:ciq:xal:3 http://stix.mitre.org/XMLSchema/external/oasis_ciq_3.0/xAL.xsd
	urn:oasis:names:tc:ciq:xpil:3 http://stix.mitre.org/XMLSchema/external/oasis_ciq_3.0/xPIL.xsd
	urn:oasis:names:tc:ciq:xnl:3 http://stix.mitre.org/XMLSchema/external/oasis_ciq_3.0/xNL.xsd" id="DocID111" version="1.1.1" timestamp="2020-11-23T14:26:05">
<stix:STIX_Header>
        <stix:Title>Package Title</stix:Title>
        <stix:Package_Intent xsi:type="stixVocabs:PackageIntentVocab-1.0">Indicators</stix:Package_Intent>
        <stix:Description>This is the package overview</stix:Description>
        <stix:Handling>
            <marking:Marking>
                <marking:Controlled_Structure>//node() | //@*</marking:Controlled_Structure>
                <marking:Marking_Structure xsi:type="AIS:AISMarkingStructure">
                    <AIS:Is_Proprietary CISA_Proprietary="true">
                        <AIS:AISConsent consent="EVERYONE"/>
                        <AIS:TLPMarking color="WHITE"/>
                    </AIS:Is_Proprietary>
                </marking:Marking_Structure>
                <marking:Information_Source>
                    <stixCommon:Identity xsi:type="stix-ciqidentity:CIQIdentity3.0InstanceType">
                        <stix-ciqidentity:Specification xmlns:stix-ciqidentity="http://stix.mitre.org/extensions/Identity#CIQIdentity3.0-1">
  <xpil:PartyName xmlns:xpil="urn:oasis:names:tc:ciq:xpil:3">
    <xnl:OrganisationName xmlns:xnl="urn:oasis:names:tc:ciq:xnl:3">
      <xnl:NameElement>CISA</xnl:NameElement>
    </xnl:OrganisationName>
  </xpil:PartyName>
  <xpil:Addresses xmlns:xpil="urn:oasis:names:tc:ciq:xpil:3">
    <xpil:Address>
      <xal:Country xmlns:xal="urn:oasis:names:tc:ciq:xal:3">
        <xal:NameElement xal:NameCode="US" xal:NameCodeType="ISO 3166-1 alpha-2"/>
      </xal:Country>
      <xal:AdministrativeArea xmlns:xal="urn:oasis:names:tc:ciq:xal:3">
        <xal:NameElement xal:NameCode="US-DC" xal:NameCodeType="ISO 3166-2"/>
      </xal:AdministrativeArea>
    </xpil:Address>
  </xpil:Addresses>
  <xpil:OrganisationInfo xmlns:xpil="urn:oasis:names:tc:ciq:xpil:3" xpil:IndustryType="Government Facilities Sector"/>
</stix-ciqidentity:Specification>
                    </stixCommon:Identity>
                </marking:Information_Source>
            </marking:Marking>
        </stix:Handling>
        <stix:Information_Source>
            <stixCommon:Contributing_Sources>
                <stixCommon:Source><stixCommon:Identity xsi:type="stix-ciqidentity:CIQIdentity3.0InstanceType">
                        <stix-ciqidentity:Specification xmlns:stix-ciqidentity="http://stix.mitre.org/extensions/Identity#CIQIdentity3.0-1">
  <xpil:PartyName xmlns:xpil="urn:oasis:names:tc:ciq:xpil:3">
    <xnl:OrganisationName xmlns:xnl="urn:oasis:names:tc:ciq:xnl:3">
      <xnl:NameElement>CISA</xnl:NameElement>
    </xnl:OrganisationName>
  </xpil:PartyName>
  <xpil:Addresses xmlns:xpil="urn:oasis:names:tc:ciq:xpil:3">
    <xpil:Address>
      <xal:Country xmlns:xal="urn:oasis:names:tc:ciq:xal:3">
        <xal:NameElement xal:NameCode="US" xal:NameCodeType="ISO 3166-1 alpha-2"/>
      </xal:Country>
      <xal:AdministrativeArea xmlns:xal="urn:oasis:names:tc:ciq:xal:3">
        <xal:NameElement xal:NameCode="US-DC" xal:NameCodeType="ISO 3166-2"/>
      </xal:AdministrativeArea>
    </xpil:Address>
  </xpil:Addresses>
  <xpil:OrganisationInfo xmlns:xpil="urn:oasis:names:tc:ciq:xpil:3" xpil:IndustryType="Government Facilities Sector"/>
</stix-ciqidentity:Specification>
                    </stixCommon:Identity>
                    <stixCommon:Tools>
                        <cyboxCommon:Tool>
                            <cyboxCommon:Metadata>
<cyboxCommon:Value>CISA</cyboxCommon:Value>
                            </cyboxCommon:Metadata>
                        </cyboxCommon:Tool>
                    </stixCommon:Tools>
                </stixCommon:Source>
            </stixCommon:Contributing_Sources>
            <stixCommon:Time>
                <cyboxCommon:Produced_Time>2018-04-20T00:00:00</cyboxCommon:Produced_Time>
            </stixCommon:Time>
        </stix:Information_Source>
    </stix:STIX_Header>
    <stix:Indicators>
        <stix:Indicator id="CISA:Indicator-a35dbed6-9b9c-442e-aaff-4aeeb88a43ef" timestamp="2020-11-23T19:26:05.255205+00:00" xsi:type='indicator:IndicatorType'>
            <indicator:Type xsi:type="stixVocabs:IndicatorTypeVocab-1.1">Anonymization</indicator:Type>
            <indicator:Description>This is the E-Mail indicator</indicator:Description>
            <indicator:Observable id="CISA:Observable-698a4ac8-c252-4886-9fad-c60152f229b6">
                <cybox:Object id="CISA:Object-2a5463ba-f9ea-4024-8295-7e6e7101dacd">
                    <cybox:Properties xsi:type="FileObj:FileObjectType">
                        <FileObj:File_Name condition="Equals">TheFileName.exe</FileObj:File_Name>
                        <FileObj:File_Path condition="Equals">/usr/local/bin</FileObj:File_Path>
                        <FileObj:Size_In_Bytes condition="Equals">123456</FileObj:Size_In_Bytes>
                        <FileObj:Hashes>
                            <cyboxCommon:Hash>
                                <cyboxCommon:Type condition="Equals" xsi:type="cyboxVocabs:HashNameVocab-1.0">MD5</cyboxCommon:Type>
                                <cyboxCommon:Simple_Hash_Value condition="Equals">d41d8cd98f00b204e9800998ecf8427e</cyboxCommon:Simple_Hash_Value>
                            </cyboxCommon:Hash>
                        </FileObj:Hashes>
                    </cybox:Properties>
                </cybox:Object>
            </indicator:Observable>
            <indicator:Kill_Chain_Phases>
                <stixCommon:Kill_Chain_Phase name="Delivery" phase_id="stix:KillChainPhase-79a0e041-9d5f-49bb-ada4-8322622b162d" kill_chain_id="stix:KillChain-af3e707f-2fb9-49e5-8c37-14026ca0a5ff"/>
            </indicator:Kill_Chain_Phases>
            <indicator:Sightings sightings_count="1">
                <indicator:Sighting timestamp="2018-03-13T20:42:24"/>
            </indicator:Sightings>
        </stix:Indicator>
    </stix:Indicators>
    <stix:TTPs/>
</stix:STIX_Package>
```

stacktrace:
```
C:\Users\user1\Envs\stix-elevator\Scripts\python.exe C:/Users/user1/example_run.py
[stixmarx.utils] [WARNING] [2020-11-23 15:42:57,324] No module named 'stix.report'
Results produced by the stix2-elevator are not for production purposes.
[stix2elevator.options] [201] [WARNING] [2020-11-23 15:42:57,934] Results produced by the stix2-elevator may generate warning messages which should be investigated.
[stix2elevator.options] [605] [WARNING] [2020-11-23 15:42:57,934] Malformed id DocID111. Generated a new uuid
[stix2elevator.options] [902] [INFO   ] [2020-11-23 15:42:57,936] Using parent object timestamp on CIQIdentity3_0Instance
[stix2elevator.options] [716] [WARNING] [2020-11-23 15:42:57,937] Based on CIQ information, identity--f7a9bcea-64f7-495b-8536-668acf11fe9f is assumed to be an organization
[stix2elevator.options] [902] [INFO   ] [2020-11-23 15:42:57,937] Using parent object timestamp on Address
[stix2elevator.options] [803] [WARNING] [2020-11-23 15:42:57,938] Location location--cf0f82e4-c2b6-4988-b870-4bfc9376dfb3 may not contain all aspects of the STIX 1.x address object
[stix2elevator.options] [902] [INFO   ] [2020-11-23 15:42:57,938] Using parent object timestamp on AISMarkingStructure
[stix2elevator.options] [902] [INFO   ] [2020-11-23 15:42:57,938] Using parent object timestamp on CIQIdentity3_0Instance
[stix2elevator.options] [716] [WARNING] [2020-11-23 15:42:57,938] Based on CIQ information, identity--c2876a1e-6e69-48c2-bd90-7e0c007c5c06 is assumed to be an organization
[stix2elevator.options] [212] [INFO   ] [2020-11-23 15:42:57,938] Created Marking Structure for AISMarkingStructure
[stix2elevator.options] [902] [INFO   ] [2020-11-23 15:42:57,940] Using parent object timestamp on AISMarkingStructure
[stix2elevator.options] [902] [INFO   ] [2020-11-23 15:42:57,940] Using parent object timestamp on CIQIdentity3_0Instance
[stix2elevator.options] [716] [WARNING] [2020-11-23 15:42:57,941] Based on CIQ information, identity--48f28c18-0138-43c3-a684-af7cef029dc7 is assumed to be an organization
[stix2elevator.options] [212] [INFO   ] [2020-11-23 15:42:57,941] Created Marking Structure for AISMarkingStructure
[stix2elevator.options] [625] [INFO   ] [2020-11-23 15:42:57,941] Found same marking structure MarkingSpecification, using marking-definition--51bc945c-1738-44a7-a83d-abdd54d9134a
[stix2elevator.options] [903] [WARNING] [2020-11-23 15:42:57,945] No valid time position information available in indicator--a35dbed6-9b9c-442e-aaff-4aeeb88a43ef, using parent timestamp
[stix2elevator.options] [632] [WARNING] [2020-11-23 15:42:57,945] Unknown phase_id stix:KillChainPhase-79a0e041-9d5f-49bb-ada4-8322622b162d in indicator--a35dbed6-9b9c-442e-aaff-4aeeb88a43ef
Traceback (most recent call last):
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2patterns\v21\grammars\STIXPatternParser.py", line 1266, in propTest
    la_ = self._interp.adaptivePredict(self._input,15,self._ctx)
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\antlr4\atn\ParserATNSimulator.py", line 342, in adaptivePredict
    alt = self.execATN(dfa, s0, input, index, outerContext)
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\antlr4\atn\ParserATNSimulator.py", line 414, in execATN
    raise e
antlr4.error.Errors.NoViableAltException: None

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:/Users/user1/example_run.py", line 19, in <module>
    results = elevate(r"..\idioms-xml\test.xml")
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2elevator\__init__.py", line 139, in elevate
    validation_results = validate_stix2_string(json_string, validator_options, fn or bundle_id[0])
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2elevator\utils.py", line 275, in validate_stix2_string
    results = validate_string(json_string, validator_options)
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2validator\validator.py", line 496, in validate_string
    return validate(stream, options)
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2validator\validator.py", line 431, in validate
    results = validate_parsed_json(obj_json, options)
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2validator\validator.py", line 408, in validate_parsed_json
    results = validate_instance(obj_json, options)
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2validator\validator.py", line 876, in validate_instance
    warnings = [pretty_error(x, options.verbose) for x in warnings]
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2validator\validator.py", line 876, in <listcomp>
    warnings = [pretty_error(x, options.verbose) for x in warnings]
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2validator\validator.py", line 72, in _iter_errors_custom
    for err in _iter_errors_custom(obj, checks, options):
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2validator\validator.py", line 62, in _iter_errors_custom
    for x in result:
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2validator\v21\shoulds.py", line 1135, in types_strict
    p = Pattern(pattern)
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2patterns\v21\pattern.py", line 21, in __init__
    self.__parse_tree = self.__do_parse(pattern_str)
  File "C:\Users\user1\Envs\stix-elevator\lib\site-packages\stix2patterns\v21\pattern.py", line 113, in __do_parse
    real_exc)
  File "<string>", line 3, in raise_from
stix2patterns.exceptions.ParseException: 1:59: no viable alternative at input '['

Process finished with exit code 1
```

**Comment by rpiazza** [comment]
@emmanvg - yes that pattern expression is malformed.  I'm looking into it.

**Comment by rpiazza** [comment]
@emmanvg - part of the problem is that the file part property is being used incorrectly.  The way I read https://cybox.mitre.org/language/version2.1/xsddocs/objects/File_Object.html, it should be terminated with a /

On the other hand - all of the file conditional expressions should be within the same [].

**Comment by emmanvg** [comment]
As an update, after correcting the FileObj:File_Path termination based on the CybOX docs and addressing the last comment from @rpiazza. This is how its looking:

`[file:hashes.MD5 = 'd41d8cd98f00b204e9800998ecf8427e' AND file:size = 123456] AND ([file:name = 'TheFileName.exe'] AND [directory:path = '/usr/local/bin'])`

**Comment by rpiazza** [comment]
When this issue is fixed, the (correct) pattern will be:

       [file:hashes.MD5 = 'd41d8cd98f00b204e9800998ecf8427e' AND (file:name = 'TheFileName.exe' AND 
        file:parent_directory_ref.path = '/usr/local/bin') AND file:size = 123456]

**Comment by rpiazza** [comment]
closed via 0d55db14a
